### ISSUE TITLE
Evil: running new Ivy command after tab-completion in evil-ex returns "Wrong type argument: stringp, nil"

### ISSUE THREAD

**Issue by jenseich** [abo-abo/swiper#977]
1. Install evil and turn on evil-mode.
(require 'evil) (evil-mode 1)
2. Create files for common prefix tab-completion.
   $ touch abc abcde
3. Enter evil-ex, tab-complete "abc" and select it. The ":" key in the following key sequence enters evil-ex.
 :e abc\<TAB\>\<RET\>\<RET\>
4. \<M-x\> or any following command using ivy returns: Wrong type argument: stringp, nil
5. The error is transient. Any ivy command issued after step 4 will work again, unless a sequence like in step 3 precedes it.

Backtrace of the error using counsel-M-x in step 4 (also confirmed reproducible with simple M-x):

> Debugger entered--Lisp error: (wrong-type-argument stringp nil)
evil-ex-update(5 6 0)
read-from-minibuffer("M-x " "^" (keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) keymap (3 keymap (19 . ivy-rotate-sort) (1 . ivy-toggle-ignore) (15 . ivy-occur)) (67108903 . ivy-avy) (33554464 . ivy-restrict-to-matches) (15 . hydra-ivy/body) (22 . ivy-scroll-up-command) (7 . minibuffer-keyboard-quit) (32 . self-insert-command) (18 . ivy-reverse-i-search) (19 . ivy-next-line-or-history) (remap keymap (describe-mode . ivy-help) (kill-ring-save . ivy-kill-ring-save) (kill-line . ivy-kill-line) (scroll-down-command . ivy-scroll-down-command) (scroll-up-command . ivy-scroll-up-command) (end-of-buffer . ivy-end-of-buffer) (beginning-of-buffer . ivy-beginning-of-buffer) (kill-word . ivy-kill-word) (forward-char . ivy-forward-char) (delete-char . ivy-delete-char) (backward-kill-word . ivy-backward-kill-word) (backward-delete-char-untabify . ivy-backward-delete-char) (delete-backward-char . ivy-backward-delete-char) (previous-line . ivy-previous-line) (next-line . ivy-next-line)) (9 . ivy-partial-or-done) (10 . ivy-alt-done) (27 keymap (1 . ivy-read-action) (15 . ivy-dispatching-call) (111 . ivy-dispatching-done) (105 . ivy-insert-current) (106 . ivy-yank-word) (114 . ivy-toggle-regexp-quote) (16 . ivy-previous-line-and-call) (14 . ivy-next-line-and-call) (118 . ivy-scroll-down-command) (112 . ivy-previous-history-element) (110 . ivy-next-history-element) (10 . ivy-immediate-done) (13 . ivy-call)) (13 . ivy-done)) nil extended-command-history)
ivy-read("M-x " [content-type-p \ Apply\ and\ Save\ new-fprs highlight-numbers-mode-hook viper-util qp-or-base64 ox-man debugger-previous-window verilog-mode-indent org-export-org custom-comment-hide rfc822-snarf-frob-list method-not-allowed :insert-button-args unexpected nomesg nnir custom-state-buffer-message doc-view-set-slice-using-mouse custom-face-state-set-and-redraw custom-magic-value-create gnus-article-highlight cal-html tweak :custom-level erc-mode-line-and-header custom-browse-group-tag custom-group-visibility message-simplify-subject-functions gnus-score-expire its-part-1 its-part-2 shr-use-fonts apropos-regexp srecode/expandproto addr custom-group-tag-faces em-script custom-unlispify-tag-names makefile-delete-backslash org-export-pdf epg--status-BADARMOR mm-use-ultra-safe-encoding custom-reset-button-menu old-groups Encrypt\ Message cancel-messages ps-print-printer custom-face-get-current-spec message-insert-importance-low ...] :predicate commandp :require-match t :history extended-command-history :action #[257 "\304\305!\203\f�\306\307�!!\210��\307�!\211��\310\307�!\311")\207" [current-prefix-arg prefix-arg this-command real-this-command featurep smex smex-rank intern command-execute record] 4 "\n\n(fn CMD)"] :sort t :keymap (keymap (67108908 . counsel--info-lookup-symbol) (67108910 . counsel-find-symbol)) :initial-input "^" :caller counsel-M-x)
counsel-M-x()
funcall-interactively(counsel-M-x)
call-interactively(counsel-M-x nil nil)
command-execute(counsel-M-x)

**Comment by wkf** [comment]
Hey @jenseich, I believe this is an issue with how evil-ex's `minibuffer-exit-hook` interacts with ivy's use of recursive minibuffers. 

After hitting tab in evil-ex, we get a recursive minibuffer. After exiting this minibuffer (by selecting a completion with ivy), evil-ex's `minibuffer-exit-hook` fires. This hook does at least two things: it removes `evil-ex-teardown` from the list of `minibuffer-exit-hooks` and it removes `evil-ex-update` from the list of `after-change-functions`. Unfortunately, since `evil-ex-update` hook is buffer local, and `minibuffer-exit-hook` fires in the context of the recursive buffer, it doesn't get removed. Furthermore, since `evil-ex-teardown` has now been removed from the `minibuffer-exit-hook` list, it doesn't fire again when we exit the actual `evil-ex` minibuffer.

Anyway, I have a workaround that seems to fix the issue for me:

```
(defvar user/-evil-ex-minibuffer nil)

(defun user/-before-evil-ex-setup ()
  (setq user/-evil-ex-minibuffer (current-buffer)))

(defun user/-around-evil-ex-teardown (f)
  (when (eq user/-evil-ex-minibuffer (current-buffer))
    (funcall f)
    (setq user/-evil-ex-minibuffer nil)))

(advice-add 'evil-ex-setup :before 'user/-before-evil-ex-setup)
(advice-add 'evil-ex-teardown :around 'user/-around-evil-ex-teardown)
```

Basically, I just take note of the buffer when we call setup, and make sure we're in the buffer when teardown gets called. 

Hope that helps!

**Comment by basil-conto** [comment]
Can this issue be closed?

**Comment by abo-abo** [comment]
Closing. If it's still an issue, open a new one.
